pallets/delegation/src/lib.rs

```diff
@@ -87,47 +88,87 @@ pub mod pallet {
 	pub struct Pallet<T>(_);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			migrations::StorageMigrator::<T>::new()
+				.pre_migration()
+				.map_err(|_| "pre-migration failed")
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			migrations::StorageMigrator::<T>::new().migrate()
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade() -> Result<(), &'static str> {
+			migrations::StorageMigrator::<T>::new()
+				.post_migration()
+				.map_err(|_| "post-migration failed")
+		}
+	}
 
-	/// Delegation root nodes stored on chain.
+	/// Delegation root nodes stored on chain. DEPRECATED.
 	///
 	/// It maps from a root node ID to the full root node.
 	#[pallet::storage]
 	#[pallet::getter(fn roots)]
-	pub type Roots<T> = StorageMap<_, Blake2_128Concat, DelegationNodeIdOf<T>, DelegationRoot<T>>;
+	pub type Roots<T> =
+		StorageMap<_, Blake2_128Concat, DelegationNodeIdOf<T>, delegation_hierarchy::v0::DelegationRoot<T>>;
 
-	/// Delegation nodes stored on chain.
+	/// Delegation nodes stored on chain. DEPRECATED.
 	///
 	/// It maps from a node ID to the full delegation node.
 	#[pallet::storage]
 	#[pallet::getter(fn delegations)]
-	pub type Delegations<T> = StorageMap<_, Blake2_128Concat, DelegationNodeIdOf<T>, DelegationNode<T>>;
+	pub type Delegations<T> =
+		StorageMap<_, Blake2_128Concat, DelegationNodeIdOf<T>, delegation_hierarchy::v0::DelegationNode<T>>;
 
-	/// Children delegation nodes.
+	/// Children delegation nodes. DEPRECATED.
 	///
 	/// It maps from a delegation node ID, including the root node, to the list
 	/// of children nodes, sorted by time of creation.
 	#[pallet::storage]
 	#[pallet::getter(fn children)]
 	pub type Children<T> = StorageMap<_, Blake2_128Concat, DelegationNodeIdOf<T>, Vec<DelegationNodeIdOf<T>>>;
 
+	/// Contains the version of the latest runtime upgrade performed.
+	#[pallet::storage]
+	#[pallet::getter(fn last_upgrade_version)]
+	pub(crate) type LastUpgradeVersion<T> = StorageValue<_, u16, ValueQuery>;
```

Comment:
didn't checkout the rest of the code but i like [StorageVersion](https://github.com/KILTprotocol/mashnet-node/blob/c82cb51b3c8e20f13aa2f9dbd587890f5e31da97/pallets/parachain-staking/src/lib.rs#L530) more.

Reply:
Note: Soon, there will be support for a `pallet-version` key in the Cargo.toml which represents the `PalletVersion`. Guess what type it will be? Yes, `u16` 😅

https://github.com/paritytech/substrate/pull/9165

Reply:
Ahah too late that I already converted everything to enums 😂

Reply:
Please take a new look at the whole migrations file.